File: code/chap04/average_by_key_use_reducebykey.py
#!/usr/bin/python
#-----------------------------------------------------
# This program find average per key using 
# the reduceByKey() transformation.
#------------------------------------------------------
# Input Parameters:
#    none
#-------------------------------------------------------
# @author Mahmoud Parsian
#-------------------------------------------------------
from __future__ import print_function
import sys
from pyspark.sql import SparkSession


def create_spark_session():
    spark = SparkSession\
        .builder\
        .appName("average_by_key_use_aggregatebykey")\
        .getOrCreate()
    return spark
#end-def

def main():
    # create an instance of SparkSession
    spark = create_spark_session()

    input = [("k1", 1), ("k1", 2), ("k1", 3), ("k1", 4), ("k1", 5),\
             ("k2", 6), ("k2", 7), ("k2", 8),\
             ("k3", 10), ("k3", 12)]

    # build RDD<key, value>
    rdd = spark.sparkContext.parallelize(input)

    # map each (key, value) into (key, (value, 1))
    pairs = rdd.map(lambda kv: (kv[0], (kv[1], 1)))
    # pairs =  
    # [
    #  ("k1", (1, 1)), ("k1", (2, 1)), ("k1", (3, 1)), ("k1", (4, 1)), ("k1", (5, 1)),
    #  ("k2", (6, 1)), ("k2", (7, 1)), ("k2", (8, 1)), 
    #  ("k3", (10, 1)), ("k3", (12, 1))
    # ]
          
    # reduce by key: 
    # x = (sum1, count1)
    # y = (sum2, count2)
    # x + y --> (sum1+sum2, count1+count2)
    sum_count = pairs.reduceByKey(lambda x, y:  (x[0] + y[0], x[1] + y[1]))
    

    # show sum_count
    print("sum_count = ", sum_count.collect())
    # [
    #  ('k3', (22, 2)), 
    #  ('k2', (21, 3)), 
    #  ('k1', (15, 5))
    # ]

    # find averages 
    #     v = (sum-of-values, count-of-values)
    #     v[0] = sum-of-values
    #     v[1] = count-of-values
    #
    avg = sum_count.mapValues( lambda v : float(v[0]) / float(v[1]))
    print("avg = ", avg.collect())
    # avg.collect()
    # [
    #  ('k3', 11), 
    #  ('k2', 7), 
    #  ('k1', 3)
    # ]
    
    # done!
    spark.stop()
#end-def

if __name__ == '__main__':
    main()

File: code/chap01/scala/src/main/scala/org/data/algorithms/spark/ch01/AverageByKeyUseReduceByKey.scala
package org.data.algorithms.spark.ch01

import org.apache.spark.sql.SparkSession
import org.data.algorithms.spark.ch01.AverageByKeyUseFoldByKey.createPair

/**
 *-----------------------------------------------------
 * Apply a reduceByKey() transformation to an
 * RDD[(key, value)] to find average per key.
 *-------------------------------------------------------
 *
 * @author Biman Mandal
 *-------------------------------------------------------
 */
object AverageByKeyUseReduceByKey {

  def createPair(t3: (String, String, Int)): (String, (Int, Int)) = {
    // t3 = (name, city, number)
    val name = t3._1
    // city = t3._2
    val number = t3._3
    // return (k, v) pair
    // v = (sum, count)
    (name, (number, 1))
  }

  def main(args: Array[String]): Unit = {

    // create an instance of SparkSession
    val spark = SparkSession.builder().master("local[*]").getOrCreate();

    /**
     *----------------------------------------
     * reduceByKey() transformation
     *
     * source_rdd.reduceByKey(func) --> target_rdd
     *
     * RDD<K,V> reduceByKey(Function2<V,V,V> func)
     * Merge the values for each key using an associative
     * and commutative reduce function. This will also
     * perform the merging locally on each mapper before
     * sending results to a reducer, similarly to a "combiner"
     * in MapReduce. Output will be hash-partitioned with the
     * existing partitioner/ parallelism level.
     *----------------------------------------
     *
     * Create a list of tuples.
     * Each tuple contains name, city, and age.
     * Create a RDD from the list above.
     */
    val listOfTuples=
      List(("alex","Sunnyvale", 25),
        ("alex","Sunnyvale", 33),
        ("alex","Sunnyvale", 45),
        ("alex","Sunnyvale", 63),
        ("mary", "Ames", 22),
        ("mary", "Cupertino", 66),
        ("mary", "Ames", 20),
        ("bob", "Ames", 26))

    println("listOfTuples = " + listOfTuples)
    val rdd = spark.sparkContext.parallelize(listOfTuples)
    println("rdd = " +  rdd)
    println("rdd.count() = " + rdd.count())
    print("rdd.collect() = " + rdd.collect().mkString("Array(", ", ", ")"))

    /**
     * ----------------------------------------------
     *  apply a map() transformation to rdd
     *  create a (key, value) pair
     *   where
     *        key is the name (first element of tuple)
     *        value is a of tuple of (number, 1)
     * ----------------------------------------------
     */
    val rdd2 = rdd.map(createPair)
    println("rdd2 = " + rdd2)
    println("rdd2.count() = " + rdd2.count())
    println("rdd2.collect() = " + rdd2.collect().mkString("Array(", ", ", ")"))


    /**
     * ------------------------------------
     *  apply a reduceByKey() transformation to rdd2
     *  create a (key, value) pair
     *   where
     *        key is the name
     *        value is the (sum, count)
     * ------------------------------------
     *  x = (sum1, count1)
     *  y = (sum2, count2)
     *  x._1+y._1 --> sum1+sum2
     *  x._2+y._2 --> count1+count2
     *
     *  zero_value = (0, 0) = (sum, count)
     */
    val sumCount = rdd2.reduceByKey((x, y) => (x._1 + y._1, x._2 + y._2))

    println("sumCount = " + sumCount)
    println("sumCount.count() = " + sumCount.count())
    println("sumCount.collect() = " + sumCount.collect().mkString("Array(", ", ", ")"))

    // find average per key
    // sumCount = [(k1, (sum1, count1)), (k2, (sum2, count2)), ...]
    val averages = sumCount.mapValues(sumAndCount => sumAndCount._1.toFloat / sumAndCount._2.toFloat )
    println("averages = " + averages)
    println("averages.count() = " + averages.count())
    println("averages.collect() = " + averages.collect().mkString("Array(", ", ", ")"))

    // done!
    spark.stop()
  }

}
